p11-kit/export-object.c

```diff
@@ -163,28 +419,37 @@ export_object (const char *token_str)
 	}
 
 	p11_kit_iter_begin (iter, modules);
-	while (p11_kit_iter_next (iter) == CKR_OK) {
-		rv = p11_kit_iter_get_attributes (iter, &attr, 1);
-		if (rv != CKR_OK) {
-			p11_message (_("failed to retrieve attribute of an object"));
+	if (p11_kit_iter_next (iter) != CKR_OK) {
+		p11_message (_("no matching object"));
+		goto cleanup;
+	}
+
+	if (p11_kit_iter_get_attributes (iter, &attr, 1) != CKR_OK) {
+		p11_message (_("failed to retrieve attribute of an object"));
+		goto cleanup;
+	}
+
+	switch (klass) {
+	case CKO_CERTIFICATE:
+		if (!export_certificate (iter, &buf))
 			goto cleanup;
-		}
+		break;
+	case CKO_PUBLIC_KEY:
+		if (!export_pubkey (iter, &buf))
+			goto cleanup;
+	default:
```

Comment:
I would say we should return an error if we match an object that is neither a certificate not public key